framework/aster-frame/src/sync/condvar.rs

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+use core::{
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+
+use super::{Mutex, MutexGuard, WaitQueue};
+
+/// A condition variable.
+pub struct Condvar {
+    wait_queue: WaitQueue,
+}
+
+impl Condvar {
+    /// Creates a new condition variable.
+    pub const fn new() -> Self {
+        Condvar {
+            wait_queue: WaitQueue::new(),
+        }
+    }
+
+    /// Waits for the condition variable to be signaled or broadcasted.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
```

Comment:
The rust-std requires this function panics if it is used with more than one mutex, should we add the check somewhere?